### ISSUE TITLE
Over-ride pen thickness

### ISSUE THREAD

**Issue by JamesBremner** [JamesBremner/DXF_Viewer#46]
The viewer should have a menu option to over-ride the pen thickness for **all** objects.

This will be applied to objects that

 - have no pen thickness option in the DXF standard ( e.g. spline )
 - where the pen thickness specification is missing from the DXF file
 - where the pen thickness IS specified in the the DXF file.  It will be over-ridden.

**Comment by JamesBremner** [comment]
The DXF standard makes no mention of the spline thickness.

https://help.autodesk.com/view/OARX/2023/ENU/?guid=GUID-E1F884F8-AA90-4864-A215-3182D47A9C74

I will add, using code 39, same as for the LINE.

**Comment by JamesBremner** [comment]
To adjust spline line thickness add code 39 to the spline specification.  See `tid46.dxf` for an example.

To change the default thickness, edit the CSpline constructor.

https://github.com/JamesBremner/DXF_Viewer/blob/845ba4c0b61840bd6aef9e2dfa40785c579954f3/src/Spline.cpp#L38

**Comment by JamesBremner** [comment]
@asmwarrior Please test

**Comment by JamesBremner** [comment]
@asmwarrior Please test

**Comment by asmwarrior** [comment]
I have a question.

Currently only the line and the spline has the default thickness.

Do we need to set the default thickness for other kinds of graphics?

**Comment by JamesBremner** [comment]
AFAIK all geometric primitives have a default line thickness of 1.

**Comment by JamesBremner** [comment]
Can you clarify what it is you want?
Do you want to specify, in DXF-Viewer rather than the DXF file, a 'default' pen thickness, other than 1, for every geometric primitive?  For each individually, or for all together?  Do you want the thickness specified in DXF-viewer to override any thickness specified in the DXF file?

**Comment by asmwarrior** [comment]
> Can you clarify what it is you want? Do you want to specify, in DXF-Viewer rather than the DXF file, a 'default' pen thickness, other than 1, for every geometric primitive? For each individually, or for all together? Do you want the thickness specified in DXF-viewer to override any thickness specified in the DXF file?

I think if some geometric primitive have the "thickness" option, we need to use that option to draw the "lines or the curves".

In my personal project, I sometimes see the line is too thin, so I would like to set the line width to 2 or 3 globally, so in this case, I would like to set them all together. (I mainly change the code in the wxWidgets port to set the line width globally)

**Comment by JamesBremner** [comment]
You have answered the question " each individually, or for all together"  Your answer was "all together"

Please answer the question "Do you want the thickness specified in DXF-viewer to override any thickness specified in the DXF file?"

**Comment by asmwarrior** [comment]
> Please answer the question "Do you want the thickness specified in DXF-viewer to override any thickness specified in the DXF file?"

Yes, the DXF-viewer has an option to "override" the thckness specified in the DXF file. If that option is NOT on, I think we should still follow the thichness specified in the DXF file.

Thanks.

**Comment by JamesBremner** [comment]
Edited description of enhancement https://github.com/JamesBremner/DXF_Viewer/issues/46#issue-2758092637

**Comment by JamesBremner** [comment]
https://github.com/user-attachments/assets/03cdecd9-94ce-4ab5-92dc-0e1ae2e1e62b

**Comment by asmwarrior** [comment]
https://github.com/user-attachments/assets/3a24c5dd-0f79-4a6f-8a4e-b8eda89956f4

I try to implement the same feature in the wxWidgets ports, and here is the video of the screen cast.

I looked at the code, and I see that in your windex code:

~~~
    mv.append("Pen",
              [&](const std::string &title) {
                    penDlg(fm,dxf);
                    auto fname = fm.text();
                    if( fname != "DXF Viewer")
                        dxf.LoadFile( fname );
                    fm.update();
              });
~~~

This means that when I set a "global thickness", I need to "reload" the dxf file and re-parse the dxf file.

I think we can improve it by setting the "thickness" in the final "drawing" stage, not in the parsing stage. So that for a large dxf file, we can save a lot of parsing time when we change the "global thickness option".

Any ideas?

I mean, when parsing the dxf file, if the thickness is not specified in the dxf file, it should have a default value such as 1. When drawing it on the screen, we can decide whether we draw it in the global thickness option or the parsed value.

**Comment by JamesBremner** [comment]
"I need to "reload" the dxf file and re-parse the dxf file."

You do not.  It is done automatically.  Please test using the latest release, or look at the video I posted.

**Comment by asmwarrior** [comment]
I think the "pen thickness override" should only happens inside the Draw() function, but not in the parsing stage.

Here is the current code inside windex.cpp

~~~
    mv.append("Pen",
              [&](const std::string &title)
              {
                  penDlg(fm, dxf);
                  auto fname = fm.text();
                  if (fname != "DXF Viewer")
                      dxf.LoadFile(fname);
                  fm.update();
              });
~~~

When the penDlg get opened, the `dxf` object reference get passed in, and in the dialog, the `dxf.penThickOverride(thick);` function will be called, and this override thickness value get modified.

But you can see in the above code, you see `dxf.LoadFile()` get called again, which means the dxf file will get opened and reparsed. When parsing, the `override thickness value` will be used.

I think this is not a good idea. Because I think the `override thickness value` should only affect the "Draw" stage of the dxf object, so that we don't need to "reload and reparse the file".

So, in this code:

~~~
    void CDxf::Draw(int width, int height)
    {
        if( ! myGraphObject.size() )
            return;

//        raven::set::cRunWatch aWatcher(" CDxf::Draw");

        // scale to window
        myBoundingRectangle.CalcScale(
            width,
            height);

        for (auto po : myGraphObject)
            po->Draw(this);
    }
~~~

In side the `Draw()` function, it should check whether the `override thickness value` should be used.

For example, in the wxWidgets' Draw function for the CCircle class:

~~~
void CCircle::Draw( CDxf* dxf )
{
    cDrawPrimitiveData draw( dxf );

    // loop over drawing primitives
    while( getDraw( draw ) )
    {
        wxColour c(draw.color);
        // Set the brush to be transparent
        theDC->SetBrush(*wxTRANSPARENT_BRUSH);
        theDC->SetPen(wxPen(c, draw.thick));
        theDC->DrawCircle( draw.x1, draw.y1, draw.r );
    }
}
~~~

the `dxf`'s `penThickOverride()` function should be checked, if the override option is enabled, we need to use a global thickness value instead of the draw.thick.

By using this way, we don't need to "reload or reparse" the file, only "redraw" the dxf file.

I hope you can understand my idea.

**Comment by JamesBremner** [comment]
>  which means the dxf file will get opened and reparsed. 

Why do you care?  A file of 4,000 plus splines loads in 150 milliseconds.  Whether or not the file is reloaded is imperceptible to the the user.

https://github.com/JamesBremner/DXF_Viewer/issues/48#issuecomment-2574081895

**Comment by asmwarrior** [comment]
> > which means the dxf file will get opened and reparsed.
> 
> Why do you care? A file of 4,000 plus splines loads in 150 milliseconds. Whether or not the file is reloaded is imperceptible to the the user.
> 
> [#48 (comment)](https://github.com/JamesBremner/DXF_Viewer/issues/48#issuecomment-2574081895)

Well, the user(maybe my client) will have some large dxf file (for example, the dxf file size which is larger than 20Mbytes), and it certainly contains more splines. So, I think reload(reparsing) the file should be avoided.

**Comment by JamesBremner** [comment]
Let me know when you have done some timing tests on your largest file.

**Comment by asmwarrior** [comment]
I think modify that code is not complicated. I will do it myself and make a pull request.

**Comment by JamesBremner** [comment]
Please do not bother.  

Your proposal makes no user visible difference to a very obscure feature.
